### PR TITLE
Implement basic functionality

### LINKED ISSUE #10: Implement "lock" functions in engage.py
Subtask of #6

The [engage.py](https://github.com/alan-turing-institute/repro-catalogue/blob/2745de147f5cb0ba5c8e045e03c868dcbe3d1682/catalogue/engage.py) file calls functions not yet implemented anywhere. These are:
- [x] `lock`
- [x] `unlock`
- [x] `check_against_lock`

From the README: 
- `lock()` should create a `.loc` file after `catalogue engage` is called
- `unlock()` reads and deletes the `.loc` file after `catalogue disengage` is called
- `check_againats_lock()` checks the created hashes after analysis against hashes in the `.loc` file (before analysis)

### LINKED ISSUE #11: Implement hash functions
Subtask of #6 

Existing hash functions (in `catalogue.py`):
- `hash_file`
- `hash_dir_by_file`
- `hash_dir_full`
- `store_hash`
- `load_hash`

Functions that are used but not yet implemented:
- [x] `get_h`
- [x] `hash_input`
- [x] `hash_output`
- [x] `hash_code`

Most of these should just call the existing functions.

### PR DESCRIPTION
**Pull request by radka-j**
Closes #10 
Closes #11 

The basic functionality is implemented:

1. When the user calls `catalogue engage` this create a `catalogue_files/.loc` file with hash of the timestamp, the input data and the code directories. 

2. When `catalogue disengage` is called, the `.loc` file is read, the retrieved hashes are compared with new hashes of the input data and code directories. 

3. If the check passed, a `catalogue_files/<TIMESTAMP>.json` file is created with the results. 

4. The result of the hash check is printed to the console.

Questions for reviews:
- are we happy with the output format or do we want to change it?
- any other design changes/preferences?

### CODE REVIEW

File: catalogue/engage.py [hunk 1]
```diff
@@ -17,41 +24,90 @@ def git_query(*args):
...
+    input_data : str
+        Path to input data directory.
+    code : str
+        Path to analysis directory.
     """
-    pass
+    hash_dict = ct.construct_dict(timestamp, input_data, code)
+    if not os.path.exists(CATALOGUE_DIR):
+        os.makedirs(CATALOGUE_DIR)
+    with open(CATALOGUE_LOCK_PATH, "w") as f:
```
Comment (edaub) [inline catalogue/engage.py:90]: Should you check that a lock file exists here and raise an exception? Isla's readme proposes that is the behavior we want.

**Review by edaub — APPROVED** [review]
I am confused how/if the `lock` function fails if there is an existing lock file, though I may have missed something (see specific inline comment). Since this isn't critical and puts in functionality where it didn't exist before, you can feel free to merge and fix in the next update if you like.

Otherwise looks fine. I think we could be a bit more careful with error handling and input checking, but fine to get things implemented and merged, and then cleaned up later.

**Comment by radka-j** [comment]
Good point, I meant to do that and forgot! Thanks for spotting this. I will implement before merging.

I agree that a bit of refactoring/clean up start of next week would be good.

**Comment by radka-j** [comment]
I added a check for whether `.lock` file exists. If it does, `catalogue engage` will throw an error.

**Pull request merged**